### PR TITLE
chore: implement basic workspace file functionality

### PR DESCRIPTION
**Pull request by f01dab1e**
# Description

implement basic workspace file functionality

## Problem

step towards https://github.com/noir-lang/noir/issues/794

## Summary

# PR Checklist

- [ ] I have tested the changes locally.
- [ ] I have formatted the changes with [Prettier](https://prettier.io/) and/or `cargo fmt` on default settings.

### CODE REVIEW

File: crates/noirc_driver/src/lib.rs [hunk 1]
```diff
@@ -154,6 +154,8 @@ pub fn check_crate(
     deny_warnings: bool,
     enable_slices: bool,
 ) -> Result<Warnings, ErrorsAndWarnings> {
+    dbg!(());
```
Comment (kevaundray) [inline crates/noirc_driver/src/lib.rs:157]: Still in draft, so no need to fix now, noting

File: crates/nargo/src/manifest/mod.rs [hunk 2]
```diff
@@ -24,8 +28,14 @@ impl PackageManifest {
     }
 }
 
+#[derive(Default, Debug, Deserialize, Clone)]
+pub struct WorkspaceConfig {
```
Comment (kevaundray) [inline crates/nargo/src/manifest/mod.rs:32]: can you put a doc comment explaining what this is?

File: crates/nargo/src/manifest/mod.rs [hunk 3]
```diff
@@ -24,8 +28,14 @@ impl PackageManifest {
...
+    #[serde(default = "Vec::default")]
```
Comment (kevaundray) [inline crates/nargo/src/manifest/mod.rs:33]: Why is this needed?

Comment (f01dab1e) [reply crates/nargo/src/manifest/mod.rs:33]: I don't think a default is needed here

**Comment by TomAFrench** [comment]
I don't think we want a package manifest and a workspace manifest represented by the same struct. these two things are separate in my mind.

File: crates/nargo_cli/src/resolver.rs [hunk 4]
```diff
@@ -154,7 +155,7 @@ fn cache_dep(
     ) -> Result<CachedDep, DependencyResolutionError> {
         let (entry_path, crate_type) = super::lib_or_bin(dir_path)?;
         let manifest_path = super::find_package_manifest(dir_path)?;
-        let manifest = super::manifest::parse(manifest_path)?;
+        let manifest = super::manifest::parse(manifest_path)?.to_package().unwrap();
```
Comment (kevaundray) [inline crates/nargo_cli/src/resolver.rs:158]: Can you use `expect` instead of `unwrap` and say why it should never fail?

File: crates/nargo_cli/src/resolver.rs [hunk 5]
```diff
@@ -70,28 +70,29 @@ pub(crate) fn resolve_root_manifest(
...
+            let (entry_path, crate_type) = super::lib_or_bin(dir_path)?;
+            let crate_id = create_local_crate(&mut context, entry_path, crate_type);
 
-        root
+            let pkg_root = manifest_path.parent().expect("Every manifest path has a parent.");
+            resolve_manifest(&mut context, crate_id, package, pkg_root)?;
+        }
+        Manifest::Workspace(workspace) => {
+            let members = workspace.config.members;
+            let root = dir_path.join(members.last().unwrap());
```
Comment (kevaundray) [inline crates/nargo_cli/src/resolver.rs:83]: Same here, can we change from unwrap to expect with a reason

File: crates/nargo/src/manifest/mod.rs [hunk 6]
```diff
@@ -6,30 +6,48 @@ pub use self::errors::InvalidPackageError;
...
 pub struct PackageManifest {
-    #[serde(default = "WorkspaceConfig::default")]
-    pub workspace: WorkspaceConfig,
-    #[serde(default = "PackageMetadata::default")]
     pub package: PackageMetadata,
-    #[serde(default = "BTreeMap::default")]
     pub dependencies: BTreeMap<String, Dependency>,
 }
 
+#[derive(Debug, Deserialize, Clone)]
```
Comment (kevaundray) [inline crates/nargo/src/manifest/mod.rs:13]: Can you add a doc comment here explaining what a manifest is?

**Review by kevaundray — COMMENTED** [review]
Great work, left some comments

**Review by kevaundray — APPROVED** [review]
